src/object/srv_ec_aggregate.c

```diff
@@ -560,31 +806,779 @@ agg_update_vos(struct ec_agg_entry *entry, bool write_parity)
+{
+	struct ec_agg_stripe_ud	 stripe_ud = { 0 };
+	int			 rc = 0;
+	int			*status;
+
+	D_ASSERT(!write_parity ||
+		 entry->ae_sgl.sg_iovs[AGG_IOV_PARITY].iov_buf);
+
+	if (write_parity)
+		stripe_ud.asu_write_par = true;
+
+	stripe_ud.asu_agg_entry = entry;
+	rc = agg_get_obj_handle(entry);
+	if (rc) {
+		D_ERROR("Failed to open object: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+	rc = ABT_eventual_create(sizeof(*status), &stripe_ud.asu_eventual);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out;
+	}
+	rc = dss_ult_create(agg_peer_update_ult, &stripe_ud,
+			    DSS_ULT_EC, 0, 0, NULL);
+	if (rc)
+		goto ev_out;
+	rc = ABT_eventual_wait(stripe_ud.asu_eventual, (void **)&status);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto ev_out;
+	}
+	if (*status != 0)
+		rc = *status;
+ev_out:
+	ABT_eventual_free(&stripe_ud.asu_eventual);
+out:
+	return rc;
+}
+
+static void
+agg_process_holes_ult(void *arg)
+{
+	daos_iod_t		 iod = { 0 };
+	crt_endpoint_t		 tgt_ep = { 0 };
+	struct ec_agg_stripe_ud	*stripe_ud = (struct ec_agg_stripe_ud *)arg;
+	struct ec_agg_entry	*entry = stripe_ud->asu_agg_entry;
+	struct ec_agg_extent	*agg_extent, *ext_tmp;
+	struct ec_agg_param	*agg_param;
+	struct obj_ec_rep_in	*ec_rep_in = NULL;
+	struct obj_ec_rep_out	*ec_rep_out = NULL;
+	crt_rpc_t		*rpc;
+	crt_opcode_t		 opcode;
+	unsigned int		 len = entry->ae_oca->u.ec.e_len;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	unsigned long		 ss = entry->ae_cur_stripe.as_stripenum *
+					k * len;
+	unsigned int		 last_ext_end = 0;
+	unsigned int		 ext_cnt = 0;
+	unsigned int		 ext_tot_len = 0;
+	int			 rc = 0;
+
+	D_ALLOC_ARRAY(stripe_ud->asu_recxs,
+		      entry->ae_cur_stripe.as_extent_cnt + 1);
+	if (stripe_ud->asu_recxs == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+	/* Process extent list to find what to re-replicate -- build recx array
+	 */
+	d_list_for_each_entry_safe(agg_extent, ext_tmp,
+				   &entry->ae_cur_stripe.as_dextents,
+				   ae_link) {
+		if (agg_extent->ae_recx.rx_idx - ss > last_ext_end) {
+			stripe_ud->asu_recxs[ext_cnt].rx_idx =
+				ss + last_ext_end;
+			stripe_ud->asu_recxs[ext_cnt].rx_nr =
+				agg_extent->ae_recx.rx_idx -
+				ss - last_ext_end;
+			ext_tot_len +=
+			stripe_ud->asu_recxs[ext_cnt++].rx_nr;
+		}
+		last_ext_end += agg_extent->ae_recx.rx_idx +
+				agg_extent->ae_recx.rx_nr - ss;
+	}
+
+	if (last_ext_end < k * len) {
+		stripe_ud->asu_recxs[ext_cnt].rx_idx =
+			ss + last_ext_end;
+		stripe_ud->asu_recxs[ext_cnt].rx_nr = ss + k * len -
+			(k * len - last_ext_end);
+		ext_tot_len += stripe_ud->asu_recxs[ext_cnt++].rx_nr;
+	}
+	stripe_ud->asu_cell_cnt = ext_cnt;
+	iod.iod_name = entry->ae_akey;
+	iod.iod_type = DAOS_IOD_ARRAY;
+	iod.iod_size = entry->ae_rsize;
+	iod.iod_nr = ext_cnt;
+	iod.iod_recxs = stripe_ud->asu_recxs;
+	entry->ae_sgl.sg_nr = 1;
+	entry->ae_sgl.sg_iovs[AGG_IOV_DATA].iov_len = ext_cnt * ext_tot_len *
+								entry->ae_rsize;
+	/* Pull data via dsc_obj_fetch */
```

Comment:
the peer also has these information? since it is a partial update punch?  Do we really need calculate these information for the peer on leader?

Reply:
Peer parity targets are fully passive in the design.  The leader does all of the work.

Reply:
Not so sure about this, need some thoughts.  Thanks.

Reply:
The peer doesn't look at its VOS state for this object since EC aggregation only deals with objects for which the process is running on the leader.